Add Durgod Taurus K310 keyboard

dkjer opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

Add support for new keyboard: Durgod Taurus K310 (non-backlit) full-sized 104/105-key ANSI/ISO.

This PR is based off @tylert's  PR #12274, and includes additional changes to share code with the Durgod K320 as well as improvements and cleanup from @nomis.

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

This PR contains the following changes:
* Add support for Durgod Taurus K310 non-backlit full-sized 104/105-key keyboards.
* Refactoring shared durgod/k310 and durgod/k320 code into durgod/k3x0.
* Renaming default_toggle_mac_windows to chimera keymap
* Enable LTO for all durgod/k3x0 keymaps
* Jump to bootloader when the hardware reset button is pressed
* Add Durgod K3x0 LAYOUT_tkl to support qmk_configurator for K320

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

dkjer commented:
Once this code makes its way to `master`, I will submit a PR to `qmk/qmk_configurator` with matching changes: https://github.com/dkjer/qmk_configurator/tree/durgod_k3x0

nomis reviewed (changes requested):
The `info.json` label for the `` ` `` key needs fixing for the ISO layout.

I don't know how the 2 and 3 keys are supposed to be labelled. Other keyboards have `@`, `#` which are the scan codes but it means there are two keys labelled `#`. I don't know why these aren't just labelled with the numbers?

I've not reviewed the mac layouts.

tylert commented on keyboards/durgod/k3x0/keymaps/chimera/keymap.c:
```diff
@@ -0,0 +1,373 @@
+    DF_W2MBL,              // Change layout from Windows to Mac
+    DF_M2WBL               // Change layout from Mac to Windows
+};
+
```
Is it worth doing something with `NEW_SAFE_RANGE` for enum the entries that come after `SAFE_RANGE` kinda like what's done in `k3x0.h`?...
```
enum K3x0_keycodes {
    KC_TGUI = SAFE_RANGE,   // Toggle between GUI Lock or Unlock
    NEW_SAFE_RANGE
};

#undef SAFE_RANGE
#define SAFE_RANGE NEW_SAFE_RANGE
#endif /* WINLOCK_DISABLED */
```

nomis replied:
The NEW_SAFE_RANGE is for keycodes added by the keyboard, not individual keymaps

tylert replied:
Understood.

tylert reviewed (approved):
LGTM! :+1:  I very much like the look of the cleaner diagrams in the comments too.

nomis commented on keyboards/durgod/k3x0/config.h:
```diff
@@ -1,4 +1,5 @@
-/* Copyright 2021 kuenhlee and Don Kjer
+/* Copyright 2021 kuenhlee, Don Kjer, Tyler Tidman
+ * Copyright 2021 Simon Arlott
```
That's not normal for copyright notices

tylert replied:
Understood.

dkjer replied:
I don't actually see any examples in `keyboards/` of using separate Copyright lines for multiple authors.  I do see a few examples of using `,` separated names on the same line, however.

I don't have a strong preference either way, but it seems as though combining into a single line is what others have done in the `qmk` codebase

nomis replied:
Each author has independent copyright terms so they may have to be split up for changes in future years and some copyright owners have commas in their name.

tylert commented:
For reference, the VIA work for K320 was merged in https://github.com/the-via/keyboards/pull/588.

dkjer commented:
> The `info.json` label for the `` ` `` key needs fixing for the ISO layout.
> 
> I don't know how the 2 and 3 keys are supposed to be labelled. Other keyboards have `@`, `#` which are the scan codes but it means there are two keys labelled `#`. I don't know why these aren't just labelled with the numbers?
> 
> I've not reviewed the mac layouts.

I don't believe the labels really matter in the `info.json` files; I think they are just there for to make it more human readable.   We could replace the 2nd `#` with `NUHS`, and the `\\` with `NUBS` if that make mores sense?

dkjer commented:
> For reference, the VIA work for K320 was merged in [the-via/keyboards#588](https://github.com/the-via/keyboards/pull/588).

I have a VIA branch ready to add K310 support, as well: https://github.com/dkjer/keyboards/tree/durgod_k310

nomis commented:
The `info.json` file content doesn't really matter for my use; it appears to be ok to do it this way for other keyboards.

tylert commented:
> @tylert Would you mind testing https://github.com/dkjer/keyboards/blob/durgod_k310/src/durgod/k310/k310.json out on a K310 to verify the numpad section of that via keymap has the correct matrix values?

Well, the `via` QMK keymap works fine on an ANSI K310 keyboard when used as a regular keyboard but the 1.3.1 (.deb) version of VIA doesn't automatically detect the K310 and I'm not sure how to build VIA from source yet or sneak a JSON file into the installed version of VIA so I'm unable to do much on the VIA front at the moment.  (There might be a minor bug at LGUI on layer1 in the VIA keyboard repo for the ISO K320.  The VIA 1.3.1 AppImage also comletely fails to run on Debian 10.x amd64 and the .deb has quite a few rough edges too.)

I'll be doing some more verification of the rest of the QMK keymaps built from commit `e631cf618` shortly on the ANSI K310 and K320 keyboards and look at fighting with VIA again later.

dkjer commented:
> Well, the `via` QMK keymap works fine on an ANSI K310 keyboard when used as a regular keyboard but the 1.3.1 (.deb) version of VIA doesn't automatically detect the K310 and I'm not sure how to build VIA from source yet or sneak a JSON file into the installed version of VIA so I'm unable to do much on the VIA front at the moment.

On the Mac version of VIA, you can 'Import Keymap' and just give it the k310.json file from that branch.  It worked for me on a K320, but of course I couldn't check the numpad section.

tylert commented:
> > Well, the `via` QMK keymap works fine on an ANSI K310 keyboard when used as a regular keyboard but the 1.3.1 (.deb) version of VIA doesn't automatically detect the K310 and I'm not sure how to build VIA from source yet or sneak a JSON file into the installed version of VIA so I'm unable to do much on the VIA front at the moment.
> 
> On the Mac version of VIA, you can 'Import Keymap' and just give it the k310.json file from that branch. It worked for me on a K320, but of course I couldn't check the numpad section.

Ah, that worked great, thanks.  Yes, VIA recognized the via-enabled K310 with that imported JSON and the key tester and matrix tester checks out on all keys, including the numpad.

The LGUI key on layer1 shows up as `0x5d2b` just as it does on the K320.

dkjer commented:
> The LGUI key on layer1 shows up as `0x5d2b` just as it does on the K320.

I don't think there is a way around that since VIA doesn't know about our custom keycodes, besides removing the Windows-Key lock functionality from VIA.  I don't think it hurts anything, and it should be able to be overridden if someone doesn't want that functionality.

drashna commented:
> I don't think there is a way around that since VIA doesn't know about our custom keycodes, besides removing the Windows-Key lock functionality from VIA. I don't think it hurts anything, and it should be able to be overridden if someone doesn't want that functionality.

Actually, there is.  VIA supports up to 16 custom keycodes.  If you use `USER00` instead of `SAFE_RANGE` when via is enabled, there is a "customKeycodes" section that is supported for the VIA json file (if you check out the ploopy mouse/trackball you can see this).

dkjer commented:
> > I don't think there is a way around that since VIA doesn't know about our custom keycodes, besides removing the Windows-Key lock functionality from VIA. I don't think it hurts anything, and it should be able to be overridden if someone doesn't want that functionality.
> 
> Actually, there is. VIA supports up to 16 custom keycodes. If you use `USER00` instead of `SAFE_RANGE` when via is enabled, there is a "customKeycodes" section that is supported for the VIA json file (if you check out the ploopy mouse/trackball you can see this).

Nice, I’ll take a look at that.  Thanks!

drashna commented:
It's a bit messy, but: 
https://github.com/qmk/qmk_firmware/blob/ef6802ca604eb6e011498fe0bbbb69a3915e0580/keyboards/ploopyco/trackball/trackball.h#L57-L69

And the via repo: 
https://github.com/the-via/keyboards/blob/3ccde52851f22785d9eaf098cdd353feb1a6559d/src/ploopyco/trackball/trackball.json#L6-L9

dkjer commented:
Rebase off develop

Converted WinLock key to a custom VIA key.   Fixed default_mac layout.

New VIA keyboard changes to match: https://github.com/dkjer/keyboards/tree/durgod_k310

dkjer commented:
Rebasing off develop

tylert commented:
Is there anything else preventing this PR from being merged?

dkjer commented:
Rebasing off develop

fauxpark commented on keyboards/durgod/k3x0/readme.md:
```diff
@@ -0,0 +1,53 @@
+- Short Boot0 to Vdd (See below)
+- Plug In USB
+- Make a Flash Image's Backup in case you wanted to restore the Keyboard to Stock's Image:
+  - Using DFUseDemo.exe from ST's STSW-STM32080: https://www.st.com/en/development-tools/stsw-stm32080.html
```
```suggestion
```

fauxpark commented on keyboards/durgod/k3x0/readme.md:
```diff
@@ -0,0 +1,53 @@
+- Put board into Bootloader mode, using the same method as when backing up the original Firmware (above)
+- Here are a few options for performing the initial Flash:
+  - Using [QMK Toolbox](https://github.com/qmk/qmk_toolbox)
+  - Using DFuseDemo.exe if STTub30 driver is used.
```
```suggestion
```

fauxpark commented on keyboards/durgod/k3x0/k3x0.h:
```diff
@@ -0,0 +1,170 @@
+#define XXX KC_NO
+
+/* Omit row 8 if it's not physically present */
+#if MATRIX_ROWS < 8
+# define K3x0_ROW_8(X70,  X71,  X72,  X73,  X74,  X75,  X76,  X77,  X78,  X79,  X7A,  X7B,  X7C,  X7D,  X7E,  X7F)
+#else
+# define K3x0_ROW_8(X70,  X71,  X72,  X73,  X74,  X75,  X76,  X77,  X78,  X79,  X7A,  X7B,  X7C,  X7D,  X7E,  X7F) \
+    { X70,  X71,  X72,  X73,  X74,  X75,  X76,  X77,  X78,  X79,  X7A,  X7B,  X7C,  X7D,  X7E,  X7F }
+#endif
```
Not a fan of this, probably won't work right with the data-driven stuff.

dkjer replied:
I split these out into k310.h and k320.h instead

fauxpark commented on keyboards/durgod/k3x0/keymaps/chimera/keymap.c:
```diff
@@ -0,0 +1,373 @@
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    /* Keymap _WBL: Windows Base Layer (Default Layer)
+   * ┌───┐   ┌───┬───┬───┬───┐ ┌───┬───┬───┬───┐ ┌───┬───┬───┬───┐ ┌───┬───┬───┐
+   * │Esc│   │F1 │F2 │F3 │F4 │ │F5 │F6 │F7 │F8 │ │F9 │F10│F11│F12│ │PSc│Slk│Pse│
```
Don't use NBSPs for spacing.

dkjer replied:
I'm not sure what you mean, could you elaborate a bit?

fauxpark replied:
These ASCII art comments use NBSP characters rather than normal spaces. For whatever reason a lot of text editors don't classify them as whitepace characters, and the result is that the ASCII art looks slightly better even when you have the "show whitepace" setting turned on:
![image](https://user-images.githubusercontent.com/4781841/124751827-6d682000-df6a-11eb-82db-a2574d1ea27e.png)
However, it seems this is not universal anymore - Sublime Text 4 now renders NBSPs as `<0xa0>`... which messes up all the nice formatting.
![image](https://user-images.githubusercontent.com/4781841/124752048-b0c28e80-df6a-11eb-9edb-9810469196ec.png)

dkjer replied:
Gotcha, thanks for catching that

b10102016 commented:
Hi,
thanks for your effort on this.
Does this work on k310w?

dkjer commented:
It doesn't currently have support for wireless, though it might worked when
plugged in via usb

- Don


On Thu, Jul 8, 2021 at 4:16 AM Ian Tsai ***@***.***> wrote:

> Hi,
> thanks for your effort on this.
> Does this work on k310w?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qmk/qmk_firmware/pull/12314#issuecomment-876353026>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAATT7XE6E3XB2BLJIOGXFLTWWCI3ANCNFSM4ZRJ6SCA>
> .
>

b10102016 commented:
Thanks for your reply.

After asking the question, I do a little bit google search.
It seems that the k310w using a different chip that is an BLE SoC from NXP.

So I decide to replace my old mech KB's MCU with other MCU :P

Anyway, thanks for your help ~

Donald Kjer ***@***.***> 於 2021年7月10日 週六 上午8:05寫道：

> It doesn't currently have support for wireless, though it might worked when
> plugged in via usb
>
> - Don
>
>
> On Thu, Jul 8, 2021 at 4:16 AM Ian Tsai ***@***.***> wrote:
>
> > Hi,
> > thanks for your effort on this.
> > Does this work on k310w?
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/qmk/qmk_firmware/pull/12314#issuecomment-876353026>,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AAATT7XE6E3XB2BLJIOGXFLTWWCI3ANCNFSM4ZRJ6SCA
> >
> > .
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/qmk/qmk_firmware/pull/12314#issuecomment-877516730>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABRACZ7FWWMX6BDC3Z2WFZ3TW6FFVANCNFSM4ZRJ6SCA>
> .
>

drashna commented:
Thanks!

tylert commented:
Thanks again to all the awesome QMK contributors.

@dkjer I'll retest all the ANSI K320 and ANSI K310 "chimera" and "typhon" keymaps again in a few days off the latest `develop` branch and I'll give it another try after it makes it to `master` too (hopefully during the next "breaking change" release in a mere few weeks :);  Aug 28, 2021 perhaps?).  I'm not expecting any surprises but I'll keep you posted.

I might give that VIA stuff another try sometime after https://github.com/dkjer/keyboards/tree/durgod_k310 gets merged too.
